tests_e2e/tests/publish_hostname/publish_hostname.py

```diff
@@ -0,0 +1,200 @@
+                 "to check the contents of '/sys/class/net/eth0/operstate'. If the contents of this file are 'up', "
+                 "no further action is needed. If contents are 'down', that indicates the network interface is down "
+                 "and more debugging needs to be done to confirm this is not caused by the agent.\n VM: {1}\n RG: {2}"
+                 "\nSubscriptionId: {3}\nUsername: {4}\nPassword: {5}".format(last_agent_status_time,
+                                                                              self._context.vm,
+                                                                              self._context.vm.resource_group,
+                                                                              self._context.vm.subscription,
+                                                                              self._context.username,
+                                                                              self._vm_password))
+
+    def retry_ssh_if_connection_reset(self, command: str, use_sudo=False):
+        # The agent may bring the network down and back up to publish the hostname, which can reset the ssh connection.
+        # Adding retry here for connection reset.
+        retries = 3
+        while retries > 0:
+            try:
+                return self._ssh_client.run_command(command, use_sudo=use_sudo)
+            except CommandError as e:
+                retries -= 1
+                retryable = e.exit_code == 255 and "Connection reset by peer" in e.stderr
+                if not retryable or retries == 0:
+                    raise
+                log.warning("The SSH operation failed, retrying in 30 secs")
+                sleep(30)
+
+    def run(self):
+        # Add password to VM and log. This allows us to debug with serial console if necessary
+        self.add_vm_password()
+
+        # This test looks up what hostname is published to dns. Check that the tools necessary to get hostname are
+        # installed, and if not install them.
+        lookup_cmd, dns_regex = self.check_and_install_dns_tools()
+
+        # Enable agent hostname monitoring
+        log.info("Executing script update-waagent-conf to enable agent hostname monitoring")
+        result = self._ssh_client.run_command("update-waagent-conf Provisioning.MonitorHostName=y "
+                                              "Provisioning.MonitorHostNamePeriod=30", use_sudo=True)
+        log.info("Successfully enabled agent hostname monitoring config flag: {0}".format(result))
+
+        hostname_change_ctr = 0
+        # Update the hostname 3 times
+        while hostname_change_ctr < 3:
+            try:
+                hostname = "lisa-hostname-monitor-{0}".format(hostname_change_ctr)
+                log.info("Update hostname to {0}".format(hostname))
+                self.retry_ssh_if_connection_reset("hostnamectl set-hostname {0}".format(hostname), use_sudo=True)
```

Comment:
why do we need extra ssh retry here? we already have general ssh retry for run_command.

Reply:
general ssh retry for run_command doesn't include connection reset by peer. We don't want to include that specific ssh failure in our general retry, so I added it only to this test